specs/language/resources.tex

```diff
@@ -663,19 +664,237 @@
+  };
+\end{HLSL}
+
+\p The \textit{shader constants} are the fields of the class type \texttt{T}.
+They can be referenced from anywhere within the translation unit after the
+\texttt{ConstantBuffer<T>} declaration as if they were fields of the
+\texttt{ConstantBuffer<T>} class.
+
+\p For example:
+
+\begin{HLSL}
+  struct MyConstants {
+    float4 CameraPos;
+  };
+
+  ConstantBuffer<MyConstants> CB;
+
+  float4 getCameraPosition() {
+    return CB.CameraPos;
+  }
+\end{HLSL}
+
+\p The layout rules for constant buffer declared with the
+\texttt{ConstantBuffer<T>} syntax are the same as for named \texttt{cbuffer}
+declaration groups.
+
+\p \texttt{ConstantBuffer<T>} can be defined at local scope to represent a local
+reference to a constant buffer that can be associated with a global buffer when
+assigned. The reference must be resolvable to a unique global buffer declaration
+before use.
+
+\p Constant buffers cannot be implicitly nor explicitly cast from one type to
+another. This means that local \texttt{ConstantBuffer} can only be assigned from
+\texttt{ConstantBuffer} with the same template type \texttt{T}.
+
+\Sub{Constant Buffer Layout}{Resources.cnbuf.lay}
+
+\p Constant buffer is arranged like an array of 16-byte rows, or 4-component
+vectors of 32-bit elements. Shader constants are arranged into the buffer in the
+order they were declared based on following rules.
+
+\p Specific basic types are packed into the last used row of the buffer if they
+fit into the available space, starting at the next available aligned position.
+These types include:
+\begin{itemize}
+\item \textit{scalar types}
+\item \textit{vector types}
+\item \textit{matrix types} with only one row in \textit{column\_major} storage layout
+\end{itemize}
+
+\p If they do not fit the remaining space of the row, they will placed at the
+start a new 16-byte aligned row.
+
+\p Vectors are aligned by the size of a single vector element type.
+
+\p Aggregate types like arrays and structures are always 16-byte row-aligned. If
+the aggregate ends with an element that does not completely fill a row, the
+remaining space on the last row may be used by the next value.
+
+\p Individual array elements are always 16-byte row aligned.
+
+\p Matrix types with more than one row in \textit{column\_major} storage layout
+and matrix types in \textit{row\_major} storage layout in are aligned to the
+16-byte row. Each row in storage layout (each column for \textit{column\_major}
+matrix) is aligned to a 16-byte row.
+
+\p Members of structured types used in constant buffers follow these same rules.
+Because structures are always 16-byte row-aligned, the offset layout within the
+structure is independent of the particular structure instance location.
+
+\Sub{Packoffset annotations}{Resources.cnbuf.po}
+
+\p Shader constants declared in \texttt{cbuffer} declaration group can have an
+optional \texttt{packoffset} annotation added before their \texttt{;} terminal.
```

Comment:
Should we add an `If an annotation is added to one shader constant in a declaration group then it must be added to all shader constants in that group`?